lib/ahoy_activity_utils.rb

```diff
@@ -1,83 +1,92 @@
+      events.group_by { |e| e.properties['collection_id'] }.each do |cid, event|
+        timestamps = event.map { |e| [e[:time], e[:name]] }
 
-        # We figure if you are logged in and doing stuff it counts as volunteer time
-        # This could be expanded or resctricted
-        events = Ahoy::Event
-            .where(user_id: user)
-            .where("time BETWEEN ? AND ?", day.beginning_of_day, day.end_of_day)
-            .select(:id, :user_id, :time, :name, :properties)
+        minutes = total_contiguous_seconds(timestamps) / 60
 
-        unless events.empty?
-            # We group by collection first, so we can sort and sum the timestamps
-            events
-                .group_by { |e| e.properties["collection_id"] }
-                .each do |cid, event|
-                    timestamps = event.map{|e| [e[:time], e[:name]] }
+        next unless minutes.positive?
 
-                    minutes = self.total_contiguous_seconds(timestamps) / 60
-                    
-                    unless minutes <= 0
-                        begin 
-                            activity = AhoyActivitySummary.create({
-                                date: day.beginning_of_day,
-                                user_id: user,
-                                collection_id: cid,
-                                activity: 'transcribe',
-                                minutes: minutes
-                            })
-                        rescue ActiveRecord::RecordNotUnique => e
-                            puts e
-                        else
-                            puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
-                        end
-                    end
-                end
-            end
+        begin
+          activity = AhoyActivitySummary.create(
+            {
+              date: day.beginning_of_day,
+              user_id: user,
+              collection_id: cid,
+              activity: 'transcribe',
+              minutes: minutes
+            }
+          )
+        rescue ActiveRecord::RecordNotUnique => e
+          puts e
+        else
+          puts "\t\t- Collection: #{activity.collection_id} \t#{activity.minutes} minutes\n"
         end
+      end
     end
 
-    def self.total_contiguous_seconds(times_and_names, tolerance=90.minutes)
-        total_seconds = 0
-        from_time = nil
+    latest_ahoy_activity_summary = AhoyActivitySummary.order(date: :desc).first
+    keep_after_date = [AhoyActivitySummary::KEEP_AFTER.ago.to_date, latest_ahoy_activity_summary.date.yesterday].min
+
+    exclude_actions = AhoyActivitySummary::WEEKLY_TRIAL_COHORT_TARGET_ACTIONS +
+                      AhoyActivitySummary::WEEKLY_TRANSCRIBER_COHORT_TARGET_ACTIONS
+
+    Ahoy::Event.where('date < ?', keep_after_date).where.not(name: exclude_actions).destroy_all
+    Visit.left_joins(:ahoy_events).where(ahoy_events: { id: nil }).destroy_all
```

Comment:
@benwbrum  this should delete all Visits without ahoy_events. Let me know if this is good for you

Reply:
Let's add a constraint on visit deletion to only delete visits older than 2 weeks (or whatever the constant is).  We have used the Visit data to track down problems with scrapers and bots.